File: backend/src/services/fileUploadService.js
class FileUploadService {
  constructor(admin) {
    this.admin = admin;
    this.bucket = admin.storage().bucket();
  }

  async uploadDataUrl(dataUrl, uidAluno) {
    if (!dataUrl || !String(dataUrl).trim()) return null;
    if (!dataUrl.startsWith('data:')) return dataUrl;

    const split = dataUrl.indexOf(',');
    if (split < 0) throw new Error('Comprovante invalido.');

    const metadata = dataUrl.substring(5, split);
    const base64Content = dataUrl.substring(split + 1);

    const contentType = metadata.includes(';') ? metadata.substring(0, metadata.indexOf(';')) : 'application/octet-stream';
    const extensao = contentType.includes('/') ? contentType.substring(contentType.indexOf('/') + 1) : 'bin';

    const bytes = Buffer.from(base64Content, 'base64');

    const bucketName = await this.resolverBucketExistente();
    const objectName = `solicitacoes/${uidAluno}/${this.uuid()}.${extensao}`;

    const file = this.admin.storage().bucket(bucketName).file(objectName);
    await file.save(bytes, {
      metadata: { contentType }
    });

    return `https://storage.googleapis.com/${bucketName}/${objectName}`;
  }

  async resolverBucketExistente() {
    const configured = process.env.FIREBASE_STORAGE_BUCKET;
    const projectId = this.admin.app.options.projectId;

    const candidatos = [];
    if (configured && configured.trim().length) candidatos.push(configured.trim());
    if (projectId && projectId.trim().length) {
      candidatos.push(`${projectId}.firebasestorage.app`);
      candidatos.push(`${projectId}.appspot.com`);
    }

    // A API Node não tem busca direta 'get bucket list'; tentamos abrir e se falhar, continua.
    for (const nome of candidatos) {
      try {
        const bucket = this.admin.storage().bucket(nome);
        const [exists] = await bucket.exists();
        if (exists) return nome;
      } catch (_) {
        // ignore
      }
    }
    throw new Error('Bucket do Firebase Storage nao encontrado. Configure FIREBASE_STORAGE_BUCKET com um bucket valido.');
  }

  uuid() {
    // bom o suficiente para compatibilidade
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === 'x' ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }
}

module.exports = FileUploadService;

File: backend/app/services/file_upload_service.rb
# frozen_string_literal: true

require "base64"
require "google/cloud/storage"
require "securerandom"

class FileUploadService
  def upload_data_url(data_url, uid_aluno)
    return nil if data_url.blank?
    return data_url unless data_url.start_with?("data:")

    split = data_url.index(",")
    raise ArgumentError, "Comprovante invalido." unless split

    metadata = data_url[5...split]
    base64_content = data_url[(split + 1)..]
    content_type = metadata.include?(";") ? metadata.split(";").first : "application/octet-stream"
    extensao = content_type.include?("/") ? content_type.split("/").last : "bin"

    bytes = Base64.decode64(base64_content)
    bucket_name = resolver_bucket_existente
    object_name = format("solicitacoes/%s/%s.%s", uid_aluno, SecureRandom.uuid, extensao)

    storage = Google::Cloud::Storage.new(
      project_id: Rails.application.config.firebase[:project_id],
      credentials: Rails.application.config.firebase[:credentials_path]
    )
    bucket = storage.bucket(bucket_name)
    bucket.create_file(StringIO.new(bytes), object_name, content_type: content_type)

    "https://storage.googleapis.com/#{bucket_name}/#{object_name}"
  end

  private

  def resolver_bucket_existente
    cfg = Rails.application.config.firebase
    project_id = cfg[:project_id]
    candidatos = []
    candidatos << cfg[:storage_bucket] if cfg[:storage_bucket].present?
    candidatos << "#{project_id}.firebasestorage.app" if project_id.present?
    candidatos << "#{project_id}.appspot.com" if project_id.present?

    storage = Google::Cloud::Storage.new(
      project_id: project_id,
      credentials: cfg[:credentials_path]
    )

    candidatos.uniq.each do |nome|
      return nome if storage.bucket(nome)
    end

    raise ArgumentError, "Bucket do Firebase Storage nao encontrado."
  end
end
